(#47) GEServer Installer

tst-ccamp opened the pull request:
Fixes https://github.com/google/earthenterprise/issues/47

This is the initial cut of the GE Server installer. There are a few things missing:

1. PUBLISH_ROOT_VOLUME size checking
1. ASSET_ROOT and PUBLISH_ROOT volume checking
1. Script parameter parsing and updated help

I figured I would go ahead and push this up so everyone can start testing with the defaults. The above two checks are also fairly complex and really don't add much value at this stage. I'd rather for users and developers to go ahead and have a basic installer. 

This also modifies install_fusion.sh to move several of the functions to a common.sh script so that both install_server.sh and install_fusion.sh can use them. So, we also need to run some regression testing on install_fusion.sh.

bradh commented on earth_enterprise/src/installer/common.sh:
```diff
@@ -0,0 +1,198 @@
+    local test_versionid=""
+
+    if [ -f "$OS_RELEASE1" ] || [ -f "$OS_RELEASE2" ]; then
+		if [ -f "$OS_RELEASE1" ]; then
```
Uneven indentation

bradh commented on earth_enterprise/src/installer/common.sh:
```diff
@@ -0,0 +1,198 @@
+            retval=1
+        fi
+    else
+		echo -e "\nThe installer could not determine your machine's operating system."
```
Uneven indentation

bradh commented on earth_enterprise/src/installer/common.sh:
```diff
@@ -0,0 +1,198 @@
+{
+	echo -e "\nThe temp install directory specified [$1] does not exist."
+	echo -e "Please specify the path of the extracted install files or first run"
+    echo -e "scons release=1 installdir=$1 install\n"
```
Uneven indentation

bradh commented on earth_enterprise/src/installer/common.sh:
```diff
@@ -0,0 +1,198 @@
+{
+	echo -e "\nYour server [$HOSTNAME] contains an invalid hostname value which typically"
+	echo -e "indicates an automatically generated hostname that might change over time."
+    echo -e "A subsequent hostname change would cause configuration issues for the "
```
Uneven indentation

bradh commented on earth_enterprise/src/installer/common.sh:
```diff
@@ -0,0 +1,198 @@
+
+  # add user if it does not exist
+	if [ -z "$USERNAME_EXISTS" ]; then
+    mkdir -p $BASEINSTALLDIR_OPT/.users/$1
```
Uneven indentation

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -18,41 +18,49 @@
 set +x
 
-# script arguments
+. common.sh
```
This only works if you are within the installer directory. I expected it to work from at least the `src` direcctory (one level up). Ideally work it out, but at least note it.

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+    local show_user_group_recommendation=false
+
+  # TODO: Remove debug
+  echo Num args: $#
```
Yep, really remove debug.

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -159,10 +571,10 @@ setup_geserver_daemon()
   # setup geserver daemon
   # For Ubuntu
    echo "Setting up the geserver daemon...\n"
```
This produces a literal backslash-n in the output:
```
bradh@earth:~/earthenterprise/earth_enterprise/src/installer$ sudo ./install_server.sh           

Welcome to the Google Earth Enterprise Server 5.1.3 installer.
Num args: 0

Copying files from source to target directories...DONE
Setting up system links...DONE
waiting for server to start.... done
server started
Creating geserve-databases...
 create_or_update_tables 
-------------------------
 
(1 row)

 create_tables 
---------------
 
(1 row)

 run_searchdef_upsert 
----------------------
 
(1 row)

Done.
waiting for server to shut down.... done
server stopped
upgrade done
pgsql_data
/var/opt/google/pgsql/data
The PostgreSQL component is successfully installed.
Setting up the geserver daemon...\n
 Removing any system startup links for /etc/init.d/geserver ...
 Adding system startup for /etc/init.d/geserver ...
   /etc/rc0.d/K10geserver -> ../init.d/geserver
   /etc/rc1.d/K10geserver -> ../init.d/geserver
   /etc/rc6.d/K10geserver -> ../init.d/geserver
   /etc/rc2.d/S90geserver -> ../init.d/geserver
   /etc/rc3.d/S90geserver -> ../init.d/geserver
   /etc/rc4.d/S90geserver -> ../init.d/geserver
   /etc/rc5.d/S90geserver -> ../init.d/geserver
GEE Server daemon setup ... DONE
Publishroot creation succeeded.
Please restart your server to apply the change.
   # a) Start the PSQL Server 
  # b) Install GEPlaces Database
  # c) Install SearchExample Database 
stopping the pgsql instance
 Google Earth Enterprise Server could not start properly.
Congratulations! Google Earth Enterprise has been successfully installed in the following directory:
          /opt/google 

          Start Google Earth Enterprise Server with the following command:
          /etc/init.d/geserver start

```

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+
+	while [ $# -gt 0 ]
+	do
+    echo Parsing: "${1,,}"
```
Also debug.

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+# TODO: Update for GEE Server
+show_help()
+{
+	echo -e "\nUsage: \tsudo ./install_fusion.sh [-dir /tmp/fusion_os_install -ar /gevol/assets -sv /gevol/src -u fusionuser"
```
Not `./install_fusion.sh`, but rather `.install_server.sh`

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+
+	echo -e "-h \t\tHelp - display this help screen"
+	echo -e "-dir \t\tTemp Install Directory - specify the temporary install directory. Default is [$TMPINSTALLDIR]."	
+	echo -e "-u \t\tFusion User Name - the user name to use for Fusion. Default is [$GEFUSIONUSER_NAME]. \n\t\tNote: this is only used for new installations."
```
Should not be specifying fusion user here.

arjntc1 replied:
Correct -- this should be updated to take in values for:  $GEAPACHEUSER_NAME, $GEPGUSER_NAME and $GEPGUSER_NAME

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+	echo -e "-h \t\tHelp - display this help screen"
+	echo -e "-dir \t\tTemp Install Directory - specify the temporary install directory. Default is [$TMPINSTALLDIR]."	
+	echo -e "-u \t\tFusion User Name - the user name to use for Fusion. Default is [$GEFUSIONUSER_NAME]. \n\t\tNote: this is only used for new installations."
+	echo -e "-g \t\tUser Group Name - the group name to use for the Fusion user. Default is [$GROUPNAME]. \n\t\tNote: this is only used for new installations."
```
Should not be specifying fusion group here.

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+	echo -e "-dir \t\tTemp Install Directory - specify the temporary install directory. Default is [$TMPINSTALLDIR]."	
+	echo -e "-u \t\tFusion User Name - the user name to use for Fusion. Default is [$GEFUSIONUSER_NAME]. \n\t\tNote: this is only used for new installations."
+	echo -e "-g \t\tUser Group Name - the group name to use for the Fusion user. Default is [$GROUPNAME]. \n\t\tNote: this is only used for new installations."
+	echo -e "-ar \t\tAsset Root Mame - the name of the asset root volume.  Default is [$ASSET_ROOT]. \n\t\tNote: this is only used for new installations. Specify absolute paths only."
```
Do we need this for server?

bradh replied:
Also Mame should be Name. This also occurs in the install_fusion.sh script. Its not really related, but I think it would be less likely to cause merge conflicts if you fix both on this commit.

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+	echo -e "-u \t\tFusion User Name - the user name to use for Fusion. Default is [$GEFUSIONUSER_NAME]. \n\t\tNote: this is only used for new installations."
+	echo -e "-g \t\tUser Group Name - the group name to use for the Fusion user. Default is [$GROUPNAME]. \n\t\tNote: this is only used for new installations."
+	echo -e "-ar \t\tAsset Root Mame - the name of the asset root volume.  Default is [$ASSET_ROOT]. \n\t\tNote: this is only used for new installations. Specify absolute paths only."
+	echo -e "-sv \t\tSource Volume Name - the name of the source volume.  Default is [$SOURCE_VOLUME]. \n\t\tNote: this is only used for new installations. Specify absolute paths only."
```
Do we need this for server?

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+	echo -e "-g \t\tUser Group Name - the group name to use for the Fusion user. Default is [$GROUPNAME]. \n\t\tNote: this is only used for new installations."
+	echo -e "-ar \t\tAsset Root Mame - the name of the asset root volume.  Default is [$ASSET_ROOT]. \n\t\tNote: this is only used for new installations. Specify absolute paths only."
+	echo -e "-sv \t\tSource Volume Name - the name of the source volume.  Default is [$SOURCE_VOLUME]. \n\t\tNote: this is only used for new installations. Specify absolute paths only."
+	echo -e "-nobk \t\tNo Backup - do not backup the current fusion setup. Default is to backup \n\t\tthe setup before installing."
```
server, not fusion.

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+	echo -e "-sv \t\tSource Volume Name - the name of the source volume.  Default is [$SOURCE_VOLUME]. \n\t\tNote: this is only used for new installations. Specify absolute paths only."
+	echo -e "-nobk \t\tNo Backup - do not backup the current fusion setup. Default is to backup \n\t\tthe setup before installing."
+	echo -e "-nop \t\tNo Purge - do not delete the temporary install directory upon successful run of the installer."
+	echo -e "\t\tDefault is to delete the directory after successful installation."
```
Since we have to install both fusion and server, we probably don't want to purge.

arjntc1 replied:
This is still useful and should probably remain.  We don't want to assume that every server will have both server and fusion.  This gives the user an option to not remove the temporary install directory (default is to remove it).

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+	echo -e "-nobk \t\tNo Backup - do not backup the current fusion setup. Default is to backup \n\t\tthe setup before installing."
+	echo -e "-nop \t\tNo Purge - do not delete the temporary install directory upon successful run of the installer."
+	echo -e "\t\tDefault is to delete the directory after successful installation."
+	echo -e "-nostart \tDo Not Start Fusion - after install, do not start the Fusion daemon.  Default is to start the daemon."
```
server, not fusion.

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+parse_arguments()
+{
+	local parse_arguments_retval=0
+    local show_user_group_recommendation=false
```
Uneven indentation.

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+					;;
+			-ar)
+				if [ $IS_NEWINSTALL == false ]; then
+					echo -e "\nYou cannot modify the asset root using the installer because Fusion is already installed on this server."
```
Server, not fusion.

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+            -sv)
+                
+                if [ $IS_NEWINSTALL == false ]; then
+					echo -e "\nYou cannot modify the source volume using the installer because Fusion is already installed on this server."
```
Server, not fusion, if we need it.

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+                show_user_group_recommendation=true
+
+				if [ $IS_NEWINSTALL == false ]; then
+					echo -e "\nYou cannot modify the fusion user name using the installer because Fusion is already installed on this server."
```
Server, not fusion. Probably worth a global search on the file.

bradh commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -109,10 +218,322 @@ main_postinstall()
+{
+	printf "\nCopying files from source to target directories..."
+
+	kdir -p $BASEINSTALLDIR_OPT/share/doc
```
mkdir

bradh commented:
Some of that install log is obviously because I didn't do a top-level `scons install` step.

tst-lsavoie commented:
@bradh I've fixed most of your comments except for the ones about the script parameters and running from other directories.

arjntc1 commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -93,26 +202,334 @@ main_postinstall()
+  echo -e "-nostart \tDo Not Start Server - after install, do not start the server daemon.  Default is to start the daemon."
+  echo -e "-hnf \t\tHostname Force - force the installer to continue installing with a bad \n\t\thostname. Bad hostname values are [${BADHOSTNAMELIST[*]}]."
+  echo -e "-hnmf \t\tHostname Mismatch Force - force the installer to continue installing with a \n\t\tmismatched hostname.\n"
+}
```
These command-line parameters should be specific to Earth Server.

arjntc1 commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -121,22 +538,13 @@ postgres_db_config()
   echo 2 | run_as_user $GEPGUSER_NAME "/opt/google/bin/geresetpgdb upgrade"
   echo -e "upgrade done"
-  
+
```
GEPGUSER_NAME appears to be a possible command line parameter (vs. GEFusionUser)

arjntc1 commented on earth_enterprise/src/installer/install_server.sh:
```diff
@@ -192,22 +600,20 @@ modify_files()
+  sed -i "s/IA_GEAPACHE_USER/$GEAPACHEUSER_NAME/" /opt/google/gehttpd/conf/gehttpd.conf
+  sed -i "s/IA_GEPGUSER/$GEPGUSER_NAME/" /opt/google/gehttpd/conf/gehttpd.conf
+  sed -i "s/IA_GEGROUP/$GRPNAME/" /opt/google/gehttpd/conf/gehttpd.conf
 
```
$GEAPACHEUSER_NAME, $GEPGUSER_NAME and $GEPGUSER_NAME should replace the GEFusionUser, GEFusionGroup specified in parse_arguments.  The show_help method should also be updated to reflect this.

tst-ccamp commented:
@tst-lsavoie Can you sign the CLA at  https://cla.developers.google.com/ ?

tst-ccamp commented:
@tst-lsavoie I'm planning to take out most of the parameter parsing for now and work on that as a follow-up. That way this can be merged just using default parameters, which I think will work for almost everyone.

tst-ccamp commented on earth_enterprise/src/installer/common.sh:
```diff
@@ -0,0 +1,198 @@
+}
+
+check_mismatched_hostname() {
+    if [ $HOSTNAME != $HOSTNAME_F ]; then
```
I was testing this and I don't think this function works correctly. The only difference between $HOSTNAME and $HOSTNAME_F is that in $HOSTNAME it converts capital letters to lowercase letters. So, this will return false if there's an uppercase letter in the hostname and true otherwise. It says that it's supposed to test whether the hostname is set to the same as the FQDN but I don't see how it could be doing that as both variables use the hostname -f command. Any thoughts on this? @arjntc1 @asmitaw @dattapm 

```
HOSTNAME="$(hostname -f | tr [A-Z] [a-z] | $NEWLINECLEANER)"
HOSTNAME_F="$(hostname -f | $NEWLINECLEANER)"
```

tst-lsavoie commented:
I signed it!

tst-lsavoie commented:
That was to trigger the CLA check.

tst-ccamp commented:
@tst-lsavoie and @bradh I believe I've addressed all the comments now. Take a look and try it out and see what you think. 

I left in a debug line for the log to print out commands when using the run_as_user function, this makes it easier to figure out which command had a problem when errors or warnings are printed in the log. I think this would be useful for the log, especially if we're trying to debug a remote user's issue. It would be nice if we could log more things like this, I think, but that's a task for the future.

Also, there are a few warning messages that still come up in the log, but I don't think they are super important.

bradh commented:
Recommend merge.
